### PR TITLE
Use per-device timer resolution in total_time calculations

### PR DESCRIPTION
**Pull request by matrix**
Reference: [Intel OpenCL Performance Debugging Intro](https://software.intel.com/en-us/articles/performance-debugging-intro/)

### CODE REVIEW

File: src/oclHashcat.c [hunk 1]
```diff
@@ -797,7 +797,7 @@ void status_display_automat ()
     hc_clGetEventProfilingInfo (data.ocl, device_param->event, CL_PROFILING_COMMAND_START, sizeof (time_start), &time_start, NULL);
     hc_clGetEventProfilingInfo (data.ocl, device_param->event, CL_PROFILING_COMMAND_END,   sizeof (time_end),   &time_end,   NULL);
 
-    const double total_time = (time_end - time_start) / 1000000.0;
+    const double total_time = ((double)(time_end - time_start) * (double)(device_param->device_timer_resolution)) / 1000000.0;
 
     fprintf (out, "%f\t", total_time);
```
Comment (magnumripper) [inline src/oclHashcat.c:802]: Sorry if I'm totally confused here, but is this really correct? I think it isn't.

Before this patch:
- If a device with 1 ns resolution gave a `time_end - time_start` resulting in 8000, we know it took 8 microseconds +/- one nanosecond.
- If a device with 80 ns resolution gave `time_end - time_start` resulting in 8000, we know it took 8 microseconds +/- 80 nanoseconds.

But after this patch, the latter will show up like... 8000*80 == 640000 meaning 640 microseconds? That's just not right.

**Comment by matrix** [comment]
@magnumripper everything is possible, look the reference link and propose a change :)
Bye,
matrix

File: src/oclHashcat.c (hunk 1 above)
Comment (magnumripper) [inline src/oclHashcat.c:802]: Revisiting eg. https://www.khronos.org/registry/cl/sdk/1.0/docs/man/xhtml/clGetEventProfilingInfo.html I see it's a bit ambiguous but as far as I'm aware it's like I said above: The result is _already_ nanoseconds and the _resolution is only needed to know the accuracy_.

http://stackoverflow.com/questions/34276371/timer-resolution-in-opencl-profiling

**Comment by matrix** [comment]
@magnumripper from your link:

> OpenCL devices are required to correctly track time across changes in device frequency and power states. The CL_DEVICE_PROFILING_TIMER_RESOLUTION specifies the resolution of the timer i.e. the number of nanoseconds elapsed before the timer is incremented.

**Comment by magnumripper** [comment]
> @magnumripper everything is possible, look the reference link and propose a change :)

My proposition is to revert this commit. It was correct before, it's busted now.

**Comment by magnumripper** [comment]
> OpenCL devices are required to correctly track time across changes in device frequency and power states. The CL_DEVICE_PROFILING_TIMER_RESOLUTION specifies the resolution of the timer i.e. the number of nanoseconds elapsed before the timer is incremented.

You read it as "after 80 ns, counter will increase by 1". I am pretty sure the correct interpretetion is "after 80 ns, counter will increase by 80".

A timer resolution of 80 merely says you can't measure smaller things - like 40 ns. It will go from 0 to 80 to 160.

**Comment by matrix** [comment]
You're probably right, but what is the correct usage of CL_DEVICE_PROFILING_TIMER_RESOLUTION in profiling?

**Comment by magnumripper** [comment]
We don't need it at all for our purposes. It's only there so you know what accuracy you are dealing with. My Macbook's GT650M has a resolution as bad as 1000 ns, so I simply can't measure anything smaller than a microsecond. My CPU device has the full 1 ns resolution.

**Comment by matrix** [comment]
Now I agree :)
For the current usage it's not useful.
Make happy @jsteube and do a new PR when you have time please :)

**Comment by magnumripper** [comment]
@gm4tr1x you should have a "revert" button right here below. It will create a PR for revertion of this one.

**Comment by magnumripper** [comment]
And don't be sad, you made like a hundred **damn good** commits before this one ;-)

**Comment by matrix** [comment]
Why sad ? :)
I'm happy, I just learned something new, one in "OpenCL" and the revert function of github, first time use :)

Ty,
matrix

**Pull request merged**